### PR TITLE
Updates Custom Integration sample app to use Payment Methods

### PR DESCRIPTION
**Pull request by csabol-stripe**
## Summary
Updates Custom Integration sample app to use Payment Methods

## Motivation
Payment methods are the future! Integrating with Sources is no longer recommended.

## Testing
Ran the sample app

### CODE REVIEW

**Comment by csabol-stripe** [comment]
The Card w/ Manual Integration and Apple Pay example view controller don't work currently becuase we need to update the sample backend

File: Example/Custom Integration/CardManualIntegrationExampleViewController.m [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+#import <Stripe/Stripe.h>
+#import "CardManualIntegrationExampleViewController.h"
+#import "BrowseExamplesViewController.h"
+
+/**
+ This example demonstrates creating a payment with a credit/debit card using Manual Integration.
+ It creates a Payment Method using card information collected with STPPaymentCardTextField, and
+ then sends the Payment Method ID to our example backend to create and confirm the Payment Intent.
+ */
+@interface CardManualIntegrationExampleViewController () <STPPaymentCardTextFieldDelegate, UIScrollViewDelegate>
```
Comment (yuki-stripe) [inline Example/Custom Integration/CardManualIntegrationExampleViewController.m]: Let's call this `CardManualConfirmationExampleViewController` instead, in line with the terminology here https://stripe.com/docs/payments/payment-intents/quickstart#flow-manual-confirmation

File: Example/Custom Integration/CardExampleViewController.m [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+ 1. Collect user's card information via `STPPaymentCardTextField`
+ 2. Create a `PaymentIntent` on our backend (this can happen concurrently with #1)
+ 3. Confirm PaymentIntent using the `STPPaymentMethodParams` for the user's card information.
+ 4. If the user needs to go through the 3D Secure authentication flow, use `STPRedirectContext` to do so.
+ 5. When user returns to the app, or finishes the SafariVC redirect flow, `STPRedirectContext` notifies via callback
+
+ See the documentation at https://stripe.com/docs/payments/dynamic-authentication for more information
+ on using PaymentIntents for dynamic authentication.
+ */
+@interface CardExampleViewController () <STPPaymentCardTextFieldDelegate>
```
Comment (yuki-stripe) [inline Example/Custom Integration/CardExampleViewController.m]: Let's call this `CardAutomaticConfirmationViewController` in line with the terminology here https://stripe.com/docs/payments/payment-intents/quickstart#flow-automatic-confirmation

File: Example/Custom Integration/CardManualIntegrationExampleViewController.m [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                                  withPaymentIntent:paymentIntent
+                                         completion:^(NSString *clientSecret, NSError *error) {
+                                             if (error) {
+                                                 [self.delegate exampleViewController:self didFinishWithError:error];
+                                             } else {
+                                                 [[STPAPIClient sharedClient] retrievePaymentIntentWithClientSecret:clientSecret completion:^(STPPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error) {
+                                                     if (error) {
+                                                         [self.delegate exampleViewController:self didFinishWithError:error];
+                                                     } else {
+                                                         [self.delegate exampleViewController:self didFinishWithMessage:@"Payment successfully created"];
```
Comment (yuki-stripe) [inline Example/Custom Integration/CardManualIntegrationExampleViewController.m]: We'll need to tell the backend to confirm the PaymentIntent again here

File: Stripe/PublicHeaders/STPPaymentMethodCardParams.h [hunk 4]
```diff
@@ -10,13 +10,21 @@
...
 NS_ASSUME_NONNULL_BEGIN
 
 /**
  The user's card details.
  */
 @interface STPPaymentMethodCardParams : NSObject <STPFormEncodable>
 
+/**
+ A convenience initializer for creating a payment method from a card source.
+ This should be used to help with migrations to Payment mMethods from Sources.
```
Comment (yuki-stripe) [inline Stripe/PublicHeaders/STPPaymentMethodCardParams.h]:
```suggestion
 This should be used to help with migrations to Payment Methods from Sources.
```

File: Stripe/PublicHeaders/STPPaymentCardTextField.h [hunk 5]
```diff
@@ -242,6 +242,12 @@ The curent brand image displayed in the receiver.
  */
 @property (nonatomic, copy, readwrite, nonnull) STPCardParams *cardParams;
 
+/**
+ Convenience property for creating an STPPaymentMethodCardParams instance from the currently
+ entered information.
+ */
+
```
Comment (yuki-stripe) [inline Stripe/PublicHeaders/STPPaymentCardTextField.h]: blank!

File: Example/Standard Integration (Sources Only).xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -109,7 +109,7 @@
 		04823F791A6849200098400B /* Products */ = {
 			isa = PBXGroup;
 			children = (
-				04823F781A6849200098400B /* Standard Integration (Swift).app */,
+				04823F781A6849200098400B /* Standard Integration (Sources Only).app */,
```
Comment (yuki-stripe) [inline Example/Standard Integration (Sources Only).xcodeproj/project.pbxproj]: I checked out this branch but the example .xcodeproj are red & missing in the workspace - I still see the old "Custom Integration (ObjC)" and "Standard Integration (Swift)" files as well in Example/

File: Example/Custom Integration/ApplePayExampleViewController.m [hunk 7]
```diff
@@ -120,19 +120,74 @@ - (void)paymentAuthorizationViewController:(PKPaymentAuthorizationViewController
...
+
+- (void)_createAndConfirmPaymentIntentWithPaymentMethod:(STPPaymentMethod *)paymentMethod completion:(void (^)(PKPaymentAuthorizationStatus))completion {
+
+    [self.delegate createAndConfirmPaymentIntentWithAmount:@(1000)
+                                             paymentMethod:paymentMethod.stripeId
+                                                completion:^(STPBackendResult status, STPPaymentIntent *paymentIntent, NSError *error) {
+                                                    if (error) {
+                                                        self.applePayError = error;
+                                                        completion(PKPaymentAuthorizationStatusFailure);
+                                                    } else if (paymentIntent.status == STPPaymentIntentStatusRequiresAction) {
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration/ApplePayExampleViewController.m]: Oh maybe not? I just included it for completeness but I can remove

File: Example/Custom Integration/BrowseExamplesViewController.m [hunk 8]
```diff
@@ -135,51 +137,52 @@ - (void)createBackendChargeWithSource:(NSString *)sourceID completion:(STPSource
...
         return;
     }
 
-    // This asks the backend to create a PaymentIntent for us, which can then be passed to the Stripe SDK to confirm
+    // This passes the token off to our payment backend, which will then actually complete charging the card using your Stripe account's secret key
     NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
     NSURLSession *session = [NSURLSession sessionWithConfiguration:config];
 
-    NSString *urlString = [BackendBaseURL stringByAppendingPathComponent:@"create_intent"];
+    NSString *urlString = [BackendBaseURL stringByAppendingPathComponent:@"create_charge"];
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration/BrowseExamplesViewController.m]: it was at time of writing. @mshafrir-stripe updated yesterday!

File: Example/Custom Integration/BrowseExamplesViewController.m [hunk 9]
```diff
@@ -236,4 +239,32 @@ - (void)exampleViewController:(UIViewController *)controller didFinishWithError:
     });
 }
 
+- (void)performRedirectForViewController:(UIViewController *)controller
+                       withPaymentIntent:(STPPaymentIntent *)paymentIntent
+                              completion:(STPRedirectCompletionHandler)completion {
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration/BrowseExamplesViewController.m]: good thought

File: Example/Custom Integration/CardExampleViewController.m [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+/**
+ This example demonstrates using PaymentIntents to accept card payments verified using 3D Secure.
+
+ 1. Collect user's card information via `STPPaymentCardTextField`
+ 2. Create a `PaymentIntent` on our backend (this can happen concurrently with #1)
+ 3. Confirm PaymentIntent using the `STPPaymentMethodParams` for the user's card information.
+ 4. If the user needs to go through the 3D Secure authentication flow, use `STPRedirectContext` to do so.
+ 5. When user returns to the app, or finishes the SafariVC redirect flow, `STPRedirectContext` notifies via callback
+
+ See the documentation at https://stripe.com/docs/payments/dynamic-authentication for more information
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration/CardExampleViewController.m]: 👍

File: Example/Custom Integration/CardExampleViewController.m [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+ on using PaymentIntents for dynamic authentication.
+ */
+@interface CardExampleViewController () <STPPaymentCardTextFieldDelegate>
+
+- (void)updateUIForPaymentInProgress:(BOOL)paymentInProgress;
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration/CardExampleViewController.m]: not necessary (used to be and I missed in the refactor)

File: Example/Custom Integration/CardExampleViewController.m (hunk 2 above)
Comment (csabol-stripe) [reply Example/Custom Integration/CardExampleViewController.m]: 👍

File: Example/Custom Integration/CardManualIntegrationExampleViewController.m (hunk 1 above)
Comment (csabol-stripe) [reply Example/Custom Integration/CardManualIntegrationExampleViewController.m]: 👍

File: Stripe/PublicHeaders/STPPaymentCardTextField.h (hunk 5 above)
Comment (csabol-stripe) [reply Stripe/PublicHeaders/STPPaymentCardTextField.h]: oops, don't need :)

File: Example/Custom Integration/BrowseExamplesViewController.m (hunk 8 above)
[the remark this answers is not in the document]
Comment (mshafrir-stripe) [reply Example/Custom Integration/BrowseExamplesViewController.m]: First need to merge the PR :) https://github.com/stripe/example-ios-backend/pull/49

File: Example/Custom Integration/ApplePayExampleViewController.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply Example/Custom Integration/ApplePayExampleViewController.m]: I *think* Apple Pay is SCA compliant and it should always give the frictionless flow.

 "Other card-based payment methods such as Apple Pay or Google Pay already support payment flows with a built-in layer of authentication (biometric or password). These can be a great way for businesses to offer a frictionless checkout experience while meeting the new requirements." (https://stripe.com/en-US/guides/strong-customer-authentication#how-to-authenticate-a-payment)

File: Example/Custom Integration/CardAutomaticConfirmationViewController.m [hunk 12]
```diff
@@ -20,20 +20,18 @@
  4. If the user needs to go through the 3D Secure authentication flow, use `STPRedirectContext` to do so.
  5. When user returns to the app, or finishes the SafariVC redirect flow, `STPRedirectContext` notifies via callback
 
- See the documentation at https://stripe.com/docs/payments/dynamic-authentication for more information
+ See the documentation at  https://stripe.com/docs/payments/payment-intents/ios i for more information
```
Comment (yuki-stripe) [inline Example/Custom Integration/CardAutomaticConfirmationViewController.m]: typos

File: Stripe.xcworkspace/contents.xcworkspacedata [hunk 13]
```diff
@@ -8,9 +8,9 @@
       location = "group:Example/UI Examples.xcodeproj">
    </FileRef>
    <FileRef
-      location = "group:Example/Standard Integration (Swift).xcodeproj">
+      location = "group:/Users/csabol/stripe/stripe-ios/Example/Standard Integration (Sources Only).xcodeproj">
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Stripe.xcworkspace/contents.xcworkspacedata]: d'oh!

File: Example/Custom Integration (Recommended).xcodeproj/project.pbxproj [hunk 14]
```diff
@@ -16,12 +16,12 @@
 		04533F121A68814100C7E52E /* Security.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 04533F111A68814100C7E52E /* Security.framework */; settings = {ATTRIBUTES = (Required, ); }; };
 		04533F191A688A0A00C7E52E /* Constants.m in Sources */ = {isa = PBXBuildFile; fileRef = 04533F181A688A0A00C7E52E /* Constants.m */; };
 		04D076171A69C11600094431 /* Stripe.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 04533F0C1A68812D00C7E52E /* Stripe.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		366F93B0225FF2A2005CFBF6 /* README.md in Resources */ = {isa = PBXBuildFile; fileRef = 366F93AF225FF2A2005CFBF6 /* README.md */; };
```
[the remark this answers is not in the document]
Comment (csabol-stripe) [reply Example/Custom Integration (Recommended).xcodeproj/project.pbxproj]: I believe this is fine, the path changed (see deletion on lin 70)

File: Example/Custom Integration/CardAutomaticConfirmationViewController.m [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+/**
+ This example demonstrates using PaymentIntents to accept card payments verified using 3D Secure.
+
+ 1. Collect user's card information via `STPPaymentCardTextField`
+ 2. Create a `PaymentIntent` on our backend (this can happen concurrently with #1)
+ 3. Confirm PaymentIntent using the `STPPaymentMethodParams` for the user's card information.
+ 4. If the user needs to go through the 3D Secure authentication flow, use `STPRedirectContext` to do so.
+ 5. When user returns to the app, or finishes the SafariVC redirect flow, `STPRedirectContext` notifies via callback
+
+ See the documentation at https://stripe.com/docs/payments/payment-intents/ios i for more information
```
Comment (yuki-stripe) [inline Example/Custom Integration/CardAutomaticConfirmationViewController.m]:
```suggestion
 See the documentation at https://stripe.com/docs/payments/payment-intents/ios for more information
```

**Pull request merged**